src/PluginClientFactory.php

```diff
@@ -0,0 +1,48 @@
+<?php
+
+namespace Http\Client\Common;
+
+use Http\Client\HttpAsyncClient;
+use Http\Client\HttpClient;
+
+/**
+ * @author Fabien Bourigault <bourigaultfabien@gmail.com>
+ */
+final class PluginClientFactory
+{
+    /**
+     * @var callable
+     */
+    private static $factory;
+
+    /**
+     * Set the factory to use. The callable to provide must have the same arguments and return type as
+     * PluginClientFactory::createClient.
+     *
+     * @internal
+     *
+     * @param callable $factory
+     */
+    public static function setFactory(callable $factory)
+    {
+        static::$factory = $factory;
+    }
+
+    /**
+     * @param HttpClient|HttpAsyncClient $client
+     * @param Plugin[]                   $plugins
+     * @param array                      $options
+     *
+     * @see PluginClient constructor for $options details.
+     *
+     * @return PluginClient
+     */
+    public function createClient($client, array $plugins = [], array $options = [])
+    {
+        $factory = static::$factory ?: function ($client, array $plugins, array $options) {
```

Comment:
Why not simply do: 

```php
if (static::$factory) {
  return $factory($client, $plugins, $options);
}

return new PluginClient($client, $plugins, $options);
```

It would be easier to read.